src/scenes/pt.cu

```diff
@@ -183,29 +225,84 @@ CU_HD Intersection Disc::intersect(Ray const &r, Object const *obj) const {
   return {t, obj, x, normal};
 }
 
+CU_D ObjectSample Object::sample(Vec3 const &dir, TraceContext &ctx) const {
+  ObjectSample sample;
+  if (type == SPHERE) {
+    sample = sphere.sample(dir, ctx, this);
+  } else if (type == PLANE) {
+    sample = plane.sample(dir, ctx, this);
+  } else if (type == DISC) {
+    sample = disc.sample(dir, ctx, this);
+  }
+  return sample;
+}
+
+CU_D ObjectSample Sphere::sample(Vec3 const &hitx, TraceContext &ctx,
+                                 [[maybe_unused]] Object const *obj) const {
+  Vec3 p = uniformHemisphereSampling(ctx);
+  p      = onb(p, obj->tr.translation() - hitx);
+  Vec3 n = p.normalized();
+  p      = radius * p + obj->tr.translation();
+
+  float lightAngle = atan(radius / (p - hitx).norm());
+  float solidAngle = R_PI * lightAngle * lightAngle;
+
+  return {1 / solidAngle, p, n};
+}
+CU_D ObjectSample Plane::sample(Vec3 const &dir, TraceContext &ctx,
+                                [[maybe_unused]] Object const *obj) const {
+  // TODO
+  return {};
+}
+CU_D ObjectSample Disc::sample(Vec3 const &dir, TraceContext &ctx,
+                               [[maybe_unused]] Object const *obj) const {
+  // TODO
+  return {};
+}
 /**********************************
  * BSDF
  **********************************/
+CU_D Vec3 uniformHemisphereSampling(TraceContext &ctx) {
```

Comment:
This has issues
![image](https://user-images.githubusercontent.com/18191185/99996477-797ce400-2dc4-11eb-885a-c65f9b9a770e.png)